service/history/historyEngine.go

```diff
@@ -205,6 +191,49 @@ func (e *historyEngineImpl) Stop() {
 	e.shard.GetDomainCache().UnregisterDomainChangeCallback(e.shard.GetShardID())
 }
 
+func (e *historyEngineImpl) registerDomainFailoverCallback() {
+	// first set the failover callback
+	e.shard.GetDomainCache().RegisterDomainChangeCallback(
+		e.shard.GetShardID(),
+		func(prevDomain *cache.DomainCacheEntry, nextDomain *cache.DomainCacheEntry) {
+			if prevDomain.GetFailoverVersion() < nextDomain.GetFailoverVersion() &&
+				nextDomain.GetReplicationConfig().ActiveClusterName == e.currentClusterName {
+				domainID := prevDomain.GetInfo().ID
+				e.txProcessor.FailoverDomain(domainID)
+				e.timerProcessor.FailoverDomain(domainID)
+			}
+			// v1 table domain cache entry will have this version being 0
+			if nextDomain.GetNotificationVersion() > 0 {
+				e.shard.UpdateDomainNotificationVersion(nextDomain.GetNotificationVersion())
+			}
+		},
+	)
+
+	// second check whether this shard should do catch up
+	domainNotificationVersion := e.shard.GetDomainCache().GetDomainNotificationVersion()
+	shardDomainNotificationVersion := e.shard.GetDomainNotificationVersion()
+	if domainNotificationVersion > shardDomainNotificationVersion {
+		for _, domain := range e.shard.GetDomainCache().GetAllDomain() {
```

Comment:
I think this logic should live within domainCache.  RegisterDomainCallback should take the last notification version and have logic to fire notifications since then for catch up.

Reply:
i thought the domain cache notification aims to be generic?
if so, then we should not put this domain failover specific logic in the domain cache.